### PR TITLE
Implement purchaseProductForUser and restorePurchasesForUser methods

### PR DESCRIPTION
**Pull request by RalfNieuwenhuizen**
Support giving a username to purchase and restore methods to prevent mixing up double accounts in the same apple ID. 
https://stackoverflow.com/questions/29255568/is-there-any-way-to-know-purchase-made-by-which-itunes-account-ios/29280858#29280858

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -54,6 +54,7 @@ InAppUtils.loadProducts(products, (error, products) => {
 | currencySymbol | string  | The currency symbol, i.e. "$" or "SEK"      |
 | currencyCode   | string  | The currency code, i.e. "USD" of "SEK"      |
 | priceString    | string  | Localised string of price, i.e. "$1,234.00" |
+| priceLocale    | string  | Country code of the price, i.e. "GB" or "FR"|
```
Comment (chirag04) [inline Readme.md]: can you call it `countryCode`?

File: InAppUtils/InAppUtils.m [hunk 2]
```diff
@@ -159,6 +176,16 @@ - (void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue
     [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
 }
 
+RCT_EXPORT_METHOD(restorePurchasesForUser:(NSString *)username
+                    callback:(RCTResponseSenderBlock)callback)
+{
+    NSString *restoreRequest = @"restoreRequest";
+    _callbacks[RCTKeyForInstance(restoreRequest)] = callback;
+    if(username) {
```
Comment (chirag04) [inline InAppUtils/InAppUtils.m]: we can instead do `if (!username) { callback([error]; return }`

File: InAppUtils/InAppUtils.m [hunk 3]
```diff
@@ -233,7 +233,7 @@ - (void)productsRequest:(SKProductsRequest *)request
                                       @"currencySymbol": [item.priceLocale objectForKey:NSLocaleCurrencySymbol],
                                       @"currencyCode": [item.priceLocale objectForKey:NSLocaleCurrencyCode],
                                       @"priceString": item.priceString,
-                                      @"priceLocale": item.priceLocale.countryCode,
+                                      @"countryCode": item.priceLocale.countryCode,
```
Comment (chirag04) [inline InAppUtils/InAppUtils.m]: `objectForKey: NSLocaleCountryCode`

**Comment by chirag04** [comment]
@RalfNieuwenhuizen This is awesome and thanks for sending the PR. left some notes but mostly looks good.

File: InAppUtils/InAppUtils.m (hunk 3 above)
Comment (RalfNieuwenhuizen) [inline InAppUtils/InAppUtils.m]: Why is that better, actually?

Comment (chirag04) [inline InAppUtils/InAppUtils.m]: 1) consistency with the rest of the codebase.
2) if they decide to change the key name, using an identifier(NSLocaleCountryCode) ensures we are not breaking our code?

File: InAppUtils/InAppUtils.m [hunk 4]
```diff
@@ -159,6 +176,18 @@ - (void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue
     [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
 }
 
+RCT_EXPORT_METHOD(restorePurchasesForUser:(NSString *)username
+                    callback:(RCTResponseSenderBlock)callback)
+{
+    NSString *restoreRequest = @"restoreRequest";
+    _callbacks[RCTKeyForInstance(restoreRequest)] = callback;
+    if(!username) {
+        callback(@[@"no_username"]);
```
Comment (chirag04) [inline InAppUtils/InAppUtils.m]: @RalfNieuwenhuizen can we make it little more descriptive? "username_required" maybe? just trying to ensure that we avoid confusion.

**Comment by chirag04** [comment]
@RalfNieuwenhuizen One last nit and will merge after that. 

Might be worth doing a test on your end one more time. haha. I will take your word for it that this works :)

File: InAppUtils/InAppUtils.m (hunk 4 above)
Comment (RalfNieuwenhuizen) [inline InAppUtils/InAppUtils.m]: You're right, on its way!

**Comment by chirag04** [comment]
@RalfNieuwenhuizen good for merge on your end?

**Comment by RalfNieuwenhuizen** [comment]
Tested and working, thanks for the comments!

Op 24 aug. 2017 4:10 p.m. schreef "Chirag" <notifications@github.com>:

**Comment by chirag04** [comment]
published.

**Pull request merged**